File: src/ratelimit/algorithms/leakyBucket.ts
import { RateLimitAlgorithm, RateLimitResult } from "../core/rateLimitAlgorithm"
import { RateLimiterBackend } from "../core/rateLimiterBackend"

interface BucketState {
  water: number
  lastLeak: number
}

export class LeakyBucket implements RateLimitAlgorithm {

  constructor(
    private backend: RateLimiterBackend,
    private capacity: number,
    private leakRate: number
  ) {}

  async allowRequest(key: string): Promise<RateLimitResult> {

    const now = Date.now()

    const bucket = await this.backend.update<BucketState>(key, (state) => {

      if (!state) {
        return {
          water: 1,
          lastLeak: now
        }
      }

      const elapsed = (now - state.lastLeak) / 1000

      const leaked = elapsed * this.leakRate

      state.water = Math.max(0, state.water - leaked)

      state.lastLeak = now

      if (state.water < this.capacity) {
        state.water += 1
      }

      return state

    })

    if (bucket.water > this.capacity) {

      return {
        allowed: false,
        remaining: 0,
        retryAfter: Math.ceil(1 / this.leakRate)
      }

    }

    return {
      allowed: true,
      remaining: Math.max(0, this.capacity - Math.floor(bucket.water))
    }

  }

}

File: src/distributed/lua/leakyBucket.lua
local key = KEYS[1]

local capacity = tonumber(ARGV[1])
local leakRate = tonumber(ARGV[2])
local now = tonumber(ARGV[3])

local bucket = redis.call("HMGET", key, "water", "timestamp")

local water = tonumber(bucket[1])
local last = tonumber(bucket[2])

if water == nil then
    water = 0
    last = now
end

local leaked = (now - last) * leakRate
water = math.max(0, water - leaked)

if water >= capacity then
    return 0
end

water = water + 1

redis.call("HMSET", key,
    "water", water,
    "timestamp", now
)

redis.call("EXPIRE", key, 60)

return 1